pyqtgraph/functions.py

```diff
@@ -1219,7 +1302,7 @@ def makeARGB(data, lut=None, levels=None, scale=None, useRGBA=False, output=None
                     maxVal = xp.nextafter(maxVal, 2*maxVal)
                 rng = maxVal-minVal
                 rng = 1 if rng == 0 else rng
-                data = rescaleData(data, scale/rng, minVal, dtype=dtype)
+                data = rescaleData_dispatch(xp, data, scale/rng, minVal, dtype, nanMask is not None)
```

Comment:
You didn't also use the new function for `pyqtgraph/graphicsItems/ImageItem.py:431`. Is this a case where the new code is legitimately not appropriate? I don't see how that's the case, seeing as it's nearly identical data to what `applyLookupTable`/`makeARGB` is dealing with.

I definitely agree that protecting narrow use-case optimizations is something we need to do in this library. Your new version of the code should be the current first choice of any user trying to rescale data. It's completely functionally equivalent, so far as I can tell, and generally performs better. Trying to preemptively protect the code from later pessimizations is still a good idea, but let's do it while still giving the users a best first choice.

Long-term, I think the right way to protect ( here and elsewhere ) is with continuous benchmarking.

Reply:
So I guess more explicitly, I recommend you take over the `rescaleData` function with `rescaleData_dispatch` ( probably also restoring the signature and just copying the `xp = ...` code ).

Reply:
>So I guess more explicitly, I recommend you take over the rescaleData function with rescaleData_dispatch

I have taken over ```rescaleData()``` and made ```_rescaleData_nditer()``` a private helper function.

>You didn't also use the new function for pyqtgraph/graphicsItems/ImageItem.py:431

That's the case where only 256 elements get scaled. But due to the above change, all uses of rescaling within pyqtgraph now go through the same ```rescaleData()```.